app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,240 @@
+    }
+
+    /**
+     * Returns the network list based on the user preference.
+     * If isAddAllNetwork preference is selected then returns the list of available networks.
+     * else return the current active network.
+     */
+    fun getNetworkList(): LinkedHashSet<Network> {
+        val networks: LinkedHashSet<Network> = linkedSetOf()
+        if (persistentState.isAddAllNetworks) {
+            networks.addAll(currentNetworks)
+        } else {
+            currentActiveNetwork?.let { networks.add(it) }
+        }
+        return networks
+    }
+
+    private fun handleNetworkChange(isForceUpdate : Boolean) {
+        val message = constructMessage(if (persistentState.isAddAllNetworks) MSG_ADD_ALL_NETWORKS else MSG_ADD_ACTIVE_NETWORK, isForceUpdate)
+        serviceHandler?.removeMessages(MSG_ADD_ACTIVE_NETWORK, null)
+        serviceHandler?.removeMessages(MSG_ADD_ALL_NETWORKS, null)
+        serviceHandler?.sendMessage(message)
+    }
+
+    /**
+     * Constructs the message object for Network handler.
+     * Add the active network to the message object
+     * in case of setUnderlying network has only active networks.
+     */
+    private fun constructMessage(what: Int, isForceUpdate: Boolean): Message {
+        val message = Message.obtain()
+        message?.what = what
+        message?.obj = isForceUpdate
+        return message!!
+    }
+
+    // Handles the network messages from the callback from the connectivity manager
+    private class NetworkRequestHandler(looper: Looper?, val connectionMonitor: ConnectionMonitor) : Handler(looper!!) {
+        override fun handleMessage(msg: Message) {
+            // isForceUpdate - true if onUserPreferenceChanged is changes, the messages should be
+            // processed forcefully regardless of the current and new networks.
+            val isForceUpdate : Boolean = msg.obj as Boolean
+            when(msg.what) {
+                MSG_ADD_ACTIVE_NETWORK -> processActiveNetwork(isForceUpdate)
+                MSG_ADD_ALL_NETWORKS -> processAllNetworks(isForceUpdate)
+            }
+        }
+
+        /**
+         * Handles the active network.
+         * Get active network from connectivity-manager
+         * Set the vpn underlying network when the current active network is different from
+         * already assigned active network or if there is difference in the available network set
+         */
+        private fun processActiveNetwork(isForceUpdate : Boolean){
+            val newActiveNetwork: Network? = connectionMonitor.connectivityManager.activeNetwork
+            val isNewNetwork = connectionMonitor.currentActiveNetwork?.networkHandle != newActiveNetwork?.networkHandle
+            if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - ${connectionMonitor.connectivityManager.getNetworkInfo(newActiveNetwork)?.typeName.toString()}, $isNewNetwork")
+            connectionMonitor.currentActiveNetwork = newActiveNetwork
+            connectionMonitor.currentNetworks.clear()
```

Comment:
Now that we control preference changes with `isForceUpdate`, the right thing to do with `currentNetworks` would be, for the sake of consistency, to add just the `currentActiveNetwork` to it.

```
currentNetworks = createActiveNetworksSet(currentActiveNetwork)
```

May be, we do not even need a `currentActiveNetwork` variable at all, because position `0` is always going to be the active network.

Reply:
Made the changes.

Reply:
This change was NOT made like suggested. Instead
```
currentNetworks.clear()
newActiveNetwork?.let { it -> currentNetworks.add(it) }
```

which is NOT what was suggested. Please clearly state how you're fixing it. If it isn't done in the EXACT same way as suggested then MENTION it.

Reply:
Modified the logic for the `processActiveNetwork` and `processAllNetworks`